From 453b79502dc0a0120db5db4d6c3af7b5eb5f67fa Mon Sep 17 00:00:00 2001
From: Vlad Yasevich <vyasevic@redhat.com>
Date: Wed, 10 Sep 2014 15:16:21 -0400
Subject: [net] bridge: Don't include NDA_VLAN for FDB entries with vid 0

Message-id: <1410362181-9327-9-git-send-email-vyasevic@redhat.com>
Patchwork-id: 93481
O-Subject: [RHEL7 PATCH 8/8] bridge: Don't include NDA_VLAN for FDB entries with vid 0
Bugzilla: 1139809
RH-Acked-by: Jiri Pirko <jpirko@redhat.com>
RH-Acked-by: Jiri Benc <jbenc@redhat.com>
RH-Acked-by: Flavio Leitner <fbl@redhat.com>

Bugzilla: 1139809
Brew: https://brewweb.devel.redhat.com/taskinfo?taskID=7936510

commit 47fab41ab51a18a2e5fc4ec63f16b4c6702809b6
From: Toshiaki Makita <makita.toshiaki@lab.ntt.co.jp>
Date: Wed, 30 Jul 2014 13:31:51 +0900

    bridge: Don't include NDA_VLAN for FDB entries with vid 0

    An FDB entry with vlan_id 0 doesn't mean it is used in vlan 0, but used when
    vlan_filtering is disabled.

    There is inconsistency around NDA_VLAN whose payload is 0 - even if we add
    an entry by RTM_NEWNEIGH without any NDA_VLAN, and even though adding an
    entry with NDA_VLAN 0 is prohibited, we get an entry with NDA_VLAN 0 by
    RTM_GETNEIGH.

    Dumping an FDB entry with vlan_id 0 shouldn't include NDA_VLAN.

    Signed-off-by: Toshiaki Makita <makita.toshiaki@lab.ntt.co.jp>
    Signed-off-by: David S. Miller <davem@davemloft.net>

Signed-off-by: Vladislav Yasevich <vyasevic@redhat.com>
Signed-off-by: Jiri Benc <jbenc@redhat.com>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/net/bridge/br_fdb.c b/net/bridge/br_fdb.c
index 2d1ca38..83b627d 100644
--- a/net/bridge/br_fdb.c
+++ b/net/bridge/br_fdb.c
@@ -580,7 +580,7 @@ static int fdb_fill_info(struct sk_buff *skb, const struct net_bridge *br,
  if (nla_put(skb, NDA_CACHEINFO, sizeof(ci), &ci))
   goto nla_put_failure;
 
- if (nla_put(skb, NDA_VLAN, sizeof(u16), &fdb->vlan_id))
+ if (fdb->vlan_id && nla_put(skb, NDA_VLAN, sizeof(u16), &fdb->vlan_id))
   goto nla_put_failure;
 
  return nlmsg_end(skb, nlh);
-- 
1.7.1